kate/recovery/src/commitments.rs

```diff
@@ -98,44 +96,134 @@ pub fn verify_equality(
 	commitments: &[u8],
 	rows: &[Option<Vec<u8>>],
 	index: &index::AppDataIndex,
-	dimension: &matrix::Dimensions,
+	dimensions: &matrix::Dimensions,
 	app_id: u32,
-) -> Result<bool, Error> {
-	if commitments.len() / config::COMMITMENT_SIZE != rows.len() {
-		return Err(Error::InvalidData(DataError::RowAndCommitmentsMismatch));
-	}
-
+) -> Result<(Vec<u32>, Vec<u32>), Error> {
 	if commitments.len() % config::COMMITMENT_SIZE > 0 {
 		return Err(Error::InvalidData(DataError::BadCommitmentsData));
 	}
 
-	let app_rows = com::app_specific_rows(index, dimension, app_id);
+	if <u32>::try_from(commitments.len() / config::COMMITMENT_SIZE)? != dimensions.extended_rows() {
+		return Err(Error::InvalidData(DataError::BadCommitmentsData));
+	}
 
-	let all_rows_present = rows
-		.iter()
-		.zip(0u32..)
-		.filter(|(row, _)| row.is_some())
-		.zip(app_rows.iter())
-		.all(|((_, a), &b)| a == b);
+	let app_rows = com::app_specific_rows(index, dimensions, app_id);
 
-	if !all_rows_present {
-		return Err(Error::InvalidData(DataError::RowAndCommitmentsMismatch));
+	if <u32>::try_from(rows.len())? != dimensions.extended_rows() {
+		return Ok((vec![], app_rows));
 	}
 
-	let (prover_key, _) = public_params.trim(dimension.cols() as usize)?;
-	let domain = EvaluationDomain::new(dimension.cols() as usize)?;
+	let (prover_key, _) = public_params.trim(dimensions.cols() as usize)?;
+	let domain = EvaluationDomain::new(dimensions.cols() as usize)?;
 
 	// This is a single-threaded implementation.
 	// At some point we should benchmark and decide
 	// if we need parallel commitments verification.
 	let verifications = commitments
 		.chunks_exact(config::COMMITMENT_SIZE)
 		.zip(rows.iter())
-		.map(|(commitment, row)| match row {
-			None => Ok(true),
-			Some(row) => Ok(row_commitment(&prover_key, domain, row)? == commitment),
+		.enumerate()
+		.filter(|&(index, _)| app_rows.contains(&(index as u32)))
+		.filter_map(|(index, (commitment, row))| row.as_ref().map(|row| (index, commitment, row)))
+		.map(|(index, commitment, row)| {
+			let scalars = try_into_scalars(row)?;
+			let polynomial = Evaluations::from_vec_and_domain(scalars, domain).interpolate();
+			let row_commitment = prover_key.commit(&polynomial)?.to_bytes();
+			Ok((index as u32, row_commitment == commitment))
 		})
-		.collect::<Result<Vec<bool>, Error>>()?;
+		.collect::<Result<Vec<(u32, bool)>, Error>>()?;
+
+	let verified = verifications
+		.iter()
+		.filter(|(_, is_equal)| *is_equal)
+		.map(|&(i, _)| i)
+		.collect::<Vec<u32>>();
 
-	Ok(verifications.iter().all(|&v| v))
+	let missing = app_rows
+		.into_iter()
+		.filter(|&row| !verified.contains(&row))
+		.collect::<Vec<u32>>();
```

Comment:
Remove new allocations:
```suggestion
	app_rows.retains(|&row| !verified.contains(&row));
```

Reply:
@MiguelDD1 [This](https://github.com/rust-lang/rust/issues/91497) might be an issue in regards to retain. Don't know if the subsequent change in the retain implementation changed anything, but it might be prudent to consider the perf impact mentioned in the issue. Wdyt?

Reply:
I have updated code @sh3ll3x3c , it seems that issue is solved, and vec should be smaller than 512, so lets revert if we notice anything here.